admin/server/github.go

```diff
@@ -88,53 +134,254 @@ func (s *Server) githubConnect(w http.ResponseWriter, r *http.Request, pathParam
+			_, err := s.admin.DB.UpdateUser(ctx, user.ID, user.DisplayName, user.PhotoURL, gitUser.GetLogin())
+			if err != nil {
+				s.logger.Error("failed to update user's github username")
+			}
+		}
+	}
+
+	// Verify that user installed the app on the right repo and we have access now
+	installationID, err := s.admin.GetGithubInstallation(ctx, remoteURL)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to check github repo status: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	if installationID == 0 {
+		// no access
+		// Redirect to UI retry page
+		redirectURL, err := url.Parse(s.urls.githubConnectRetry)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create retry request url: %s", err.Error()), http.StatusInternalServerError)
+			return
+		}
+
+		qry := redirectURL.Query()
+		qry.Set("remote", remoteURL)
+		redirectURL.RawQuery = qry.Encode()
+		http.Redirect(w, r, redirectURL.String(), http.StatusTemporaryRedirect)
+	}
+
 	// Redirect to UI success page
-	redirectURL, err := url.JoinPath(s.opts.FrontendURL, "/-/github/connect/success")
+	http.Redirect(w, r, s.urls.githubConnectSuccess, http.StatusTemporaryRedirect)
+}
+
+// githubAuthLogin starts user authorisation of github app.
+// In case github app is installed by another user, other users of the repo need to separately authorise github app
+// where this flow comes into picture.
+// Some implementation details are copied from auth package.
+// It's implemented as a non-gRPC endpoint mounted directly on /github/auth/login.
+func (s *Server) githubAuthLogin(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	// Generate random state for CSRF
+	b := make([]byte, 32)
+	_, err := rand.Read(b)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to generate state: %s", err), http.StatusInternalServerError)
+		return
+	}
+	state := base64.StdEncoding.EncodeToString(b)
+
+	// Get auth cookie
+	sess, err := s.cookies.Get(r, cookieName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get session: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Set state in cookie
+	sess.Values[cookieFieldState] = state
+
+	// Save cookie
+	if err := sess.Save(r, w); err != nil {
+		http.Error(w, fmt.Sprintf("failed to save session: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	oauthConf := &oauth2.Config{
+		ClientID:     s.opts.GithubClientID,
+		ClientSecret: s.opts.GithubClientSecret,
+		Endpoint:     githuboauth.Endpoint,
+		RedirectURL:  s.urls.githubAuthCallback,
+	}
+	// Redirect to github login page
+	http.Redirect(w, r, oauthConf.AuthCodeURL(state, oauth2.AccessTypeOnline), http.StatusTemporaryRedirect)
+}
+
+// githubAuthCallback is called after a user authorizes github app on their account
+// It's implemented as a non-gRPC endpoint mounted directly on /github/auth/callback.
+func (s *Server) githubAuthCallback(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
```

Comment:
Should we add an option to also check that access was granted to the correct remote here? By passing a remote to `githubAuthLogin` and checking in this callback that there's access to it. Then if not, we should redirect to the retry page instead of the success page.

Reply:
Consider the aforementioned case where someone authorizes with the wrong Github account (another one than they used to clone locally). It doesn't take much extra for us to make sure they land on a retry page.

IMO, the fact that we redirect to the user authorize page instead of the app install page is an implementation detail only due to Github's bad screen. The users will not understand the difference, we should pretend it's the same.